### ISSUE TITLE
Verbose output option (case 14634)

### ISSUE THREAD

**Issue by JKrag** [Praqma/Praqmatic-Automated-Changelog#56]
Especially for initial config & debugging, it would be very useful to have a `-v` / `--verbose` option.

Nearly all good command line tools have this, and I really could use it for PAC. 

Ideally we should have multiple levels of verbosity.

Possible suggestions for output:

* Basic "progress bar"
* commit being processed
* issues found
* Issue URL being requested

**Comment by JKrag** [comment]
I might have a go at writing and contributing this myself, possibly in increments as I find the time.

**Comment by JKrag** [comment]
I was going to write a separate issue on better error reporting, but it might actually go well with this one. PAC currently is very silent on many exceptions. It replaces them with nice groomed error messages, which is nice, but unfortunately provides no way of actually outputting the original error messages or returned http error codes.

**Comment by buep** [comment]
I only have one requirement you need to consider and document before starting implementing verbose and better error messages. That is making a strategy:

* what will verbose output?
* is there several verbose levels? if so, what includes what?
* what is the common format for output of PAC? "[PAC] .... " as I have seen somewhere?
* do error message have a specific format?

> " It replaces them with nice groomed error messages, which is nice, but unfortunately provides no way of actually outputting the original error messages or returned http error codes."
- indeed, this is not nice. I want always an interpreted error message from the developer that understand the situation, and then if it makes sense also the stack trace for easier debugging. If not as default, then available to me by a switch.


I we can come up with a description and plan for output on this issue, we can create smaller issues for each logical part and do them one by one.

**Comment by JKrag** [comment]
My plan was to implement the actual support for parsing and handling verbosity (at multiple levels), as a general feature of the tool, together with a helper method for outputting output with respect to a specified level. 

Then I would leave it up to the other developers to start making use of this in various sections when it seems relevant.

My plan is to initially include the following:
* support -v switch for verbosity. Switch can be repeated for higher verbosity (docopt should support this now).
* support -q switch for more quiet output. Also with repeat mostly for consistency.
* `verbosity = 0 - [quiet_count] + [verbose_count]`(yes, this means that you can say `-vvq` if you insist, and it is the same as `-v`. While this sounds weird, it is very standard behaviour, and useful for overriding default behaviour when combining with bash aliasing or similar). 
* I will try to integrate this as best I can with the existing handling of settings, but might need some input from @MadsNielsen on this, or leave the final cleanup to him.
* Implement a helper method for printing output. The method would have a signature like: `def verboseprint(fromlevel, *args)` 
i.e "Print this message if users selected verbosity level  is above X.
* I do not have strong opinion about whether output should be prefixed with e.g. `[PAC]`, but this should be trivially easy to add to the above mentioned verboseprint method if wanted. (note to self: there might be a catch here if we want to verboseprint multiline output)
* I am pretty confident that I can come up with some unit tests for parsing the new arguments as well. I am not sure how to test the verboseprint method, but I am up to giving it a shot (i.e. no promise) - I might learn some more ruby secrets along the way.

All told, my main plan was to provide the infrastructure to make it much easier for developers to write differentiated output. I am currently not offering to go through the entire codebase and refactor or improve the existing output.

__Please let me know if you would accept a contribution / pull-request as per the above description, then I will try to find the time to start hacking.__


P.S. yes, maybe we should at least create a separate issue on the part about improved exception handling and error reporting :-D

**Comment by buep** [comment]
Sounds like a great plan :-) If docs and some test, could just be blackbox func test is included we would appreciate it very much indeed. 

I suggest you spin another issue with your just described plan from above and link it here because it doesn't solve all your suggestions originally stated on this issue. That way we can close bits and pieces one at the time.

**Comment by MadsNielsen** [comment]
I like this idea and @JKrag fleshed it out very nicely. We just need to figure out a way to do this in a nice way as i do NOT want to pass and add an extra argument to every single method that does something :-)

**Comment by JKrag** [comment]
Don't worry @MadsNielsen. I think I have you covered. The verbosity level is stored in the core settings, and there is a nice helper method in Core that can be used for any console output, where you specify that the provided message should only be printet if the user has selected a verbosity level above X.

so standard messages could be `verboseprint(0, <message>)`while very debug-style messages could be `verboseprint(3, <message>)`.
High importance error could even be `verboseprint(-2, <message>)`.

So verbosity-compliant printing should be easy anywhere in the code where you have required Core.

**Comment by JKrag** [comment]
Oh, and I just pushed my WIP branch feature/56-verbosity so you can see it for real. I still have to find time to write more tests, and update docs before it is ready for integration. 

And maybe this shouldn't actually be in Core, but rather in a Logging module? Opinions are very welcome.

**Comment by buep** [comment]
Related core module or logging module, both are fine with me. I care about the "component" of PAC, or current module to match our architectural drawing. In your situation logging is a core thing, so split it out if you think it makes it easier and will keep files smaller and easier to read. Even though it have it's own file, I consider it part of core :-)

**Comment by buep** [comment]
Putting this in progress... there are already logged time on it, so guess it is in progress and code changes are available.

**Comment by buep** [comment]
This have been in progress very long... it is only 4 hours of work according to estimate... can I get some focus on finishing this during the week?

**Comment by buep** [comment]
Thanks, can see changes are pushed and docs and tests also :-)
